README.md
# horizontal_int_picker

A new Flutter package which is used to select integers horizontally. 

#Screenshots

<img src="https://github.com/shashankgupta3891/horizontal_int_picker/blob/master/upload_media/ss1.jpg?raw=true" width="300em" /> <img src="https://github.com/shashankgupta3891/horizontal_int_picker/blob/master/upload_media/sgif.gif?raw=true" width="300em" />

## Getting Started

This project is a starting point for a Dart
[package](https://flutter.dev/developing-packages/),
a library module containing code that can be shared easily across
multiple Flutter or Dart projects.

For help getting started with Flutter, view our 
[online documentation](https://flutter.dev/docs), which offers tutorials, 
samples, guidance on mobile development, and a full API reference.

### 🔩 Installation

Add to your [pubspec.yaml](https://github.com/shashankgupta3891/horizontal_int_picker/pubspec.yaml) :

```yaml
dependencies:
  horizontal_int_picker: <last_version>
```

### Usage 

[Example](https://github.com/shashankgupta3891/horizontal_int_picker/example)

### How to use

```dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

import 'package:horizontal_int_picker/horizontal_int_picker.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Horizontal Int Picker'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _selectedWeight = 100;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Colors.deepPurple,
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Icon(
              Icons.arrow_downward,
              size: 30,
              color: Colors.indigo,
            ),
        /// You can add this widget as a normal widget.
            HorizontalIntPicker(
              minValue: 60,
              maxValue: 120,
              initial: 90,
              selectedColor: Colors.blue,
              unitName: 'KG',
              itemSize: 90,
              onSelectedItemChanged: (val) {
                setState(() {
                  _selectedWeight = val;
                });
              },
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {},
        tooltip: 'Increment',
        child: Text(
          _selectedWeight.toString(),
          style: TextStyle(fontWeight: FontWeight.bold),
        ),
      ),
    );
  }
}

```



CHANGELOG.md
## [0.0.1] - Release

* Horizontal Integer Picker Library


lib/horizontal_int_picker.dart
library horizontal_int_picker;

import 'dart:math' as math;
import 'package:flutter/material.dart';

class HorizontalIntPicker extends StatefulWidget {
  const HorizontalIntPicker({
    this.minValue = 0,
    @required this.maxValue,
    this.initial,
    @required this.onSelectedItemChanged,
    this.selectedColor = Colors.blue,
    this.unSelectedColor = Colors.grey,
    this.unitName = '',
    this.scrollWheelDiameterRatio = 2.0,
    this.fontSize = 40,
    this.itemSize = 80.0,
  });

  final int minValue;
  final int maxValue;
  final int initial;
  final ValueChanged<int> onSelectedItemChanged;
  final Color selectedColor;
  final Color unSelectedColor;
  final String unitName;
  final double scrollWheelDiameterRatio;
  final double fontSize;
  final double itemSize;

  @override
  _HorizontalIntPickerState createState() => _HorizontalIntPickerState();
}

class _HorizontalIntPickerState extends State<HorizontalIntPicker> {
  int currentValue;
  FixedExtentScrollController fixedExtentScrollController;

  @override
  void initState() {
    super.initState();
    currentValue = widget.minValue;
    fixedExtentScrollController = FixedExtentScrollController(
        initialItem: widget.initial == null
            ? widget.minValue
            : widget.initial - widget.minValue);
  }

  @override
  Widget build(BuildContext context) {
    return Container(
      child: ClipRect(
        child: Align(
          heightFactor: widget.itemSize / MediaQuery.of(context).size.width,
          child: Container(
            height: MediaQuery.of(context).size.width,
            width: MediaQuery.of(context).size.width,
            child: Transform.rotate(
              angle: -90 * (math.pi / 180),
              child: ListWheelScrollView.useDelegate(
                controller: fixedExtentScrollController,
                diameterRatio: widget.scrollWheelDiameterRatio,
                physics: FixedExtentScrollPhysics(),
                onSelectedItemChanged: (val) {
                  setState(() {
                    currentValue = val + widget.minValue;
                  });
                  widget.onSelectedItemChanged(
                      currentValue); // Sending current value out from widget
                },
                itemExtent: widget.itemSize, //width of child
                childDelegate: ListWheelChildBuilderDelegate(
                  childCount: widget.maxValue - widget.minValue + 1,
                  builder: (context, index) {
                    int value = index + widget.minValue;

                    return Transform.rotate(
                      angle: 90 * (math.pi / 180),
                      child: Container(
                        width: widget.itemSize, //height of Child

                        child: Center(
                          child: Column(
                            mainAxisAlignment: MainAxisAlignment.spaceAround,
                            mainAxisSize: MainAxisSize.max,
                            children: [
                              Text(
                                '$value',
                                style: TextStyle(
                                  color: currentValue == value
                                      ? widget.selectedColor
                                      : widget.unSelectedColor,
                                  fontSize: widget.fontSize,
                                  fontWeight: currentValue == value
                                      ? FontWeight.bold
                                      : FontWeight.normal,
                                ),
                              ),
                              Visibility(
                                visible: currentValue == value,
                                child: Text(
                                  widget.unitName ?? '',
                                  style: TextStyle(
                                    color: currentValue == value
                                        ? widget.selectedColor
                                        : Colors.grey,
                                  ),
                                ),
                              ),
                            ],
                          ),
                        ),
                      ),
                    );
                  },
                ),
              ),
            ),
          ),
        ),
      ),
    );
  }
}


example/example.dart
import 'package:flutter/cupertino.dart';
import 'package:flutter/material.dart';

import 'package:horizontal_int_picker/horizontal_int_picker.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      debugShowCheckedModeBanner: false,
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Horizontal Int Picker'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key key, this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _selectedWeight = 100;

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        backgroundColor: Colors.deepPurple,
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Icon(
              Icons.arrow_downward,
              size: 30,
              color: Colors.indigo,
            ),
            HorizontalIntPicker(
              minValue: 60,
              maxValue: 120,
              initial: 90,
              selectedColor: Colors.blue,
              unitName: 'KG',
              itemSize: 90,
              onSelectedItemChanged: (val) {
                setState(() {
                  _selectedWeight = val;
                });
              },
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: () {},
        tooltip: 'Increment',
        child: Text(
          _selectedWeight.toString(),
          style: TextStyle(fontWeight: FontWeight.bold),
        ),
      ),
    );
  }
}
